File: RxTools-library/src/main/java/com/vondear/rxtools/view/cardstack/tools/RxStackScrollDelegateImpl.java
package com.vondear.rxtools.view.cardstack.tools;

import android.view.View;

import com.vondear.rxtools.view.cardstack.RxCardStackView;

/**
 * @author vondear
 */
public class RxStackScrollDelegateImpl implements RxScrollDelegate {

    private RxCardStackView mRxCardStackView;
    private int mScrollY;
    private int mScrollX;

    public RxStackScrollDelegateImpl(RxCardStackView rxCardStackView) {
        mRxCardStackView = rxCardStackView;
    }

    private static int clamp(int n, int my, int child) {
        if (my >= child || n < 0) {
            return 0;
        }
        if ((my + n) > child) {
            return child - my;
        }
        return n;
    }

    private void updateChildPos() {
        for (int i = 0; i < mRxCardStackView.getChildCount(); i++) {
            View view = mRxCardStackView.getChildAt(i);
            if (view.getTop() - mScrollY < mRxCardStackView.getChildAt(0).getY()) {
                view.setTranslationY(mRxCardStackView.getChildAt(0).getY() - view.getTop());
            } else if (view.getTop() - mScrollY > view.getTop()) {
                view.setTranslationY(0);
            } else {
                view.setTranslationY(-mScrollY);
            }
        }
    }

    @Override
    public void scrollViewTo(int x, int y) {
        x = clamp(x, mRxCardStackView.getWidth() - mRxCardStackView.getPaddingRight() - mRxCardStackView.getPaddingLeft(), mRxCardStackView.getWidth());
        y = clamp(y, mRxCardStackView.getShowHeight(), mRxCardStackView.getTotalLength());
        mScrollY = y;
        mScrollX = x;
        updateChildPos();
    }

    @Override
    public int getViewScrollY() {
        return mScrollY;
    }

    @Override
    public void setViewScrollY(int y) {
        scrollViewTo(mScrollX, y);
    }

    @Override
    public int getViewScrollX() {
        return mScrollX;
    }

    @Override
    public void setViewScrollX(int x) {
        scrollViewTo(x, mScrollY);
    }
}

File: RxUI/src/main/java/com/tamsiree/rxui/view/cardstack/tools/RxStackScrollDelegateImpl.kt
package com.tamsiree.rxui.view.cardstack.tools

import com.tamsiree.rxui.view.cardstack.RxCardStackView

/**
 * @author tamsiree
 * @date 2018/6/11 11:36:40 整合修改
 */
class RxStackScrollDelegateImpl(private val mRxCardStackView: RxCardStackView) : RxScrollDelegate {

    private var mScrollY = 0
    private var mScrollX = 0
    private fun updateChildPos() {
        for (i in 0 until mRxCardStackView.childCount) {
            val view = mRxCardStackView.getChildAt(i)
            if (view.top - mScrollY < mRxCardStackView.getChildAt(0).y) {
                view.translationY = mRxCardStackView.getChildAt(0).y - view.top
            } else if (view.top - mScrollY > view.top) {
                view.translationY = 0f
            } else {
                view.translationY = -mScrollY.toFloat()
            }
        }
    }

    override fun scrollViewTo(x: Int, y: Int) {
        var x = x
        var y = y
        x = clamp(x, mRxCardStackView.width - mRxCardStackView.paddingRight - mRxCardStackView.paddingLeft, mRxCardStackView.width)
        y = clamp(y, mRxCardStackView.showHeight, mRxCardStackView.totalLength)
        mScrollY = y
        mScrollX = x
        updateChildPos()
    }

    override var viewScrollY: Int
        get() = mScrollY
        set(y) {
            scrollViewTo(mScrollX, y)
        }

    override var viewScrollX: Int
        get() = mScrollX
        set(x) {
            scrollViewTo(x, mScrollY)
        }

    companion object {
        private fun clamp(n: Int, my: Int, child: Int): Int {
            if (my >= child || n < 0) {
                return 0
            }
            return if (my + n > child) {
                child - my
            } else n
        }
    }

}